### PR TITLE
i#5505 kernel tracing: Add flag to configure PT buffer size

### PR DESCRIPTION
**Pull request by abhinav92003**
Adds a new drmemtrace command-line flag that allows setting an appropriate size for the buffer used to collect PT kernel trace data during online trace collection: -kernel_trace_buffer_size_shift.

For apps with a high thread count, a large value may cause OOMs. At the same time, a too low value causes issues during trace decode. The new flag allows tuning it as needed.

Verified that the PT related tests continue to work on a system that supports the Intel-PT hardware feature:

```
The following tests passed:
	code_api|tool.drcacheoff.kernel.simple_SUDO
	code_api|tool.drcacheoff.kernel.opcode-mix_SUDO
	code_api|tool.drcacheoff.kernel.syscall-mix_SUDO
	code_api|tool.drcacheoff.kernel.invariant-checker_SUDO

...

The following tests passed:
	code_api|client.drpttracer_SUDO-test
```

Issue: #5505

### CODE REVIEW

File: clients/drcachesim/tracer/syscall_pt_trace.h [hunk 1]
```diff
@@ -198,6 +198,16 @@ class syscall_pt_trace_t {
...
      */
     file_t output_file_;
+
+    /* The ring buffer that stores the recorded PT data is assigned a size of
+     * (1 << ring_buffer_size_shift_) * page_size.
+     * For apps with a high thread count, this may cause us to exceed the available
+     * memory. But if it is configured too low, we may see errors of type
+     * READ_RING_BUFFER_ERROR_OLD_DATA_OVERWRITTEN or
+     * "get next instruction error: expected tracing enabled event" during PT trace
+     * decoding.
```
Comment (derekbruening) [inline clients/drcachesim/tracer/syscall_pt_trace.h:208]: Are these propagated to the user?  Will the user know what they mean?  Please augment the option docs to explain what error to look for to identify when the buffer is too small.  Also augment the error message returned to add a user-friendly message suggesting to increase the new option.

**Review by derekbruening — APPROVED** [review]

File: clients/drcachesim/common/options.cpp [hunk 2]
```diff
@@ -860,6 +860,12 @@ droption_t<bool> op_skip_kcore_dump(
     "as normal. This may be useful if it is not feasible to run the application "
     "with superuser permissions and the user wants to use a different kcore "
     "dump, from a prior trace or created separately.");
+droption_t<int> op_pt_ring_buffer_size_shift(
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/common/options.cpp:863]: Used op_kernel_trace_buffer_size_shift.

File: clients/drcachesim/common/options.cpp [hunk 3]
```diff
@@ -860,6 +860,12 @@ droption_t<bool> op_skip_kcore_dump(
...
+    DROPTION_SCOPE_ALL, "pt_ring_buffer_size_shift", 8,
+    "Size of the ring buffer used to collect PT trace data.",
+    "Size of the ring buffer used to collect PT trace data. The size is computed as "
+    "(1 << pt_ring_buffer_size_shift) * page_size. Too large buffers can cause OOMs "
+    "on apps with many threads.");
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/common/options.cpp:868]: Done.

File: clients/drcachesim/common/options.cpp [hunk 4]
```diff
@@ -860,6 +860,12 @@ droption_t<bool> op_skip_kcore_dump(
     "as normal. This may be useful if it is not feasible to run the application "
     "with superuser permissions and the user wants to use a different kcore "
     "dump, from a prior trace or created separately.");
+droption_t<int> op_pt_ring_buffer_size_shift(
+    DROPTION_SCOPE_ALL, "pt_ring_buffer_size_shift", 8,
+    "Size of the ring buffer used to collect PT trace data.",
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/common/options.cpp:865]: Used op_kernel_trace_buffer_size_shift.

File: clients/drcachesim/tracer/syscall_pt_trace.h (hunk 1 above)
Comment (abhinav92003) [reply clients/drcachesim/tracer/syscall_pt_trace.h:208]: Done.

File: clients/drcachesim/common/options.cpp [hunk 5]
```diff
@@ -860,6 +860,12 @@ droption_t<bool> op_skip_kcore_dump(
...
+    "Size of the ring buffer used to collect PT trace data. The size is computed as "
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/common/options.cpp]: Done.

**Pull request merged**